composer/datasets/streaming/dataset.py

```diff
@@ -0,0 +1,422 @@
+            data = fp.read(size)
+
+        return self._unpack_sample(data)
+
+    def _make_new_growing_epoch(self) -> int:
+        """Start a new growing epoch, in which we own the sample sequence because it grows.
+
+        Returns:
+            int: The epoch ID, an identifier which is given back to the caller.
+        """
+        if not self._lock:
+            raise RuntimeError("Attempted to use lock but lock was not created.")
+
+        with self._lock:
+            epoch = self._next_epoch
+            self._next_epoch += 1
+            self._epoch_to_todo_ids[epoch] = list(self._downloaded_ids)
+        return epoch
+
+    def _next_id(self, epoch: int) -> Optional[int]:
+        """Get next sample of the growing epoch given by epoch, or None if done.
+
+        If we are currently out of samples but not finished downloading the
+        shards, blocks until it has new samples.
+
+        Args:
+            epoch (int): The epoch, an identifier for this sequence of samples.
+
+        Returns:
+            int: ID of next sample.
+        """
+        if not self._lock:
+            raise RuntimeError("Attempted to use lock but lock was not created.")
+
+        while True:
+            with self._lock:
+                todo_ids = self._epoch_to_todo_ids[epoch]
+                if todo_ids:
+                    return todo_ids.pop(0)
+                elif self._are_all_shards_downloaded:
+                    del self._epoch_to_todo_ids[epoch]
+                    return None
+                else:
+                    pass
+            sleep(0.25)
+
+    def _iter_ids(self) -> Iterator[int]:
+        """Get an iterator over all our sample IDs.
+
+        Returns:
+            Iterator[int]: Each sample ID.
+        """
+        if not self._lock:
+            raise RuntimeError("Attempted to use lock but lock was not created.")
+
+        with self._lock:
+            have_full_epoch = self._are_all_shards_downloaded
+
+        if have_full_epoch:
+            ids = list(self._downloaded_ids)
+            if self.shuffle:
+                np.random.shuffle(ids)
+            for idx in ids:
+                yield idx
+        else:
+            epoch = self._make_new_growing_epoch()
+            while True:
+                idx = self._next_id(epoch)
+                if idx is None:
+                    break
+                yield idx
+
+    def __iter__(self) -> Iterator[Any]:
+        """Iterate over all the samples in our partition.
+
+        If not all samples have been downloaded yet, iterates over what it has
+        while inserting the remainder into the sequence behind the scenes as it
+        progresses.
+
+        Returns:
+            Iterator[Any]: Each sample.
+        """
+        # Lock is created here because DataLoader calls __iter__ in each worker process
+        # and the lock is worker-specific
+        if self._lock is None:
+            self._lock = Lock()
+
+        # Load samples
+        self._load()
```

Comment:
Are we sure load and lock check should go in iter? Was thinking of calling iter_ids directly for testing purposes

Reply:
This change was motivated from @ravi-mosaicml,  for all our tests we're doing end-to-end dataset build followed by `iter(dataset)` so it should be safer, if anything.